notify-debouncer-mini/src/lib.rs

```diff
@@ -253,87 +309,81 @@ impl<T: Watcher> Debouncer<T> {
     timeout: Duration,
-    tick_rate: Option<Duration>,
     mut event_handler: F,
-    config: notify::Config
+    config: notify::Config,
 ) -> Result<Debouncer<T>, Error> {
-    let data = DebounceData::default();
-
-    let stop = Arc::new(AtomicBool::new(false));
-
-    let tick_div = 4;
-    let tick = match tick_rate {
-        Some(v) => {
-            if v > timeout {
-                return Err(Error::new(ErrorKind::Generic(format!(
-                    "Invalid tick_rate, tick rate {:?} > {:?} timeout!",
-                    v, timeout
-                ))));
-            }
-            v
-        }
-        None => timeout.checked_div(tick_div).ok_or_else(|| {
-            Error::new(ErrorKind::Generic(format!(
-                "Failed to calculate tick as {:?}/{}!",
-                timeout, tick_div
-            )))
-        })?,
-    };
-
-    {
-        let mut data_w = data.lock().unwrap();
-        data_w.timeout = timeout;
-    }
+    let (tx, rx) = std::sync::mpsc::channel();
 
-    let data_c = data.clone();
-    let stop_c = stop.clone();
     let thread = std::thread::Builder::new()
         .name("notify-rs debouncer loop".to_string())
-        .spawn(move || loop {
-            if stop_c.load(Ordering::Acquire) {
-                break;
-            }
-            std::thread::sleep(tick);
-            let send_data;
-            let errors: Vec<crate::Error>;
-            {
-                let mut lock = data_c.lock().expect("Can't lock debouncer data!");
-                send_data = lock.debounced_events();
-                errors = lock.errors();
-            }
-            if send_data.len() > 0 {
-                event_handler.handle_event(Ok(send_data));
-            }
-            if errors.len() > 0 {
-                event_handler.handle_event(Err(errors));
+        .spawn(move || {
+            let mut data = DebounceDataInner::new(timeout);
+            let mut run = true;
+            while run {
+                match data.next_tick() {
+                    Some(timeout) => {
+                        dbg!(timeout);
+                        // wait for wakeup
+                        match rx.recv_timeout(timeout) {
+                            Ok(InnerEvent::NotifyEvent(e)) => data.push(e),
+                            Ok(InnerEvent::Shutdown) => run = false,
+                            Err(RecvTimeoutError::Disconnected) => run = false,
+                            // TODO: wait the correct time?
+                            Err(RecvTimeoutError::Timeout) => continue,
+                        }
+                    },
+                    None => match rx.recv() {
+                        Ok(InnerEvent::NotifyEvent(e)) => data.push(e),
+                        Ok(InnerEvent::Shutdown) => run = false,
+                        Err(_) => run = false,
+                    }
+                }
+                
+                // // drain events channel
```

Comment:
I'm not sure if we want this, it certainly allows us to receive rapid fired events.

Reply:
I don't think that it'll provide any value? The "rapid fire" events will be handled by `recv_timeout` anyway just fine.